x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BasicTokenFilter.java

```diff
@@ -0,0 +1,273 @@
+                    tokenizer,
+                    CharSeqTokenTrieNode.EMPTY,
+                    CharArraySet.EMPTY_SET,
+                    isStripAccents,
+                    isTokenizeCjkChars
+                );
+                return new TokenStreamComponents(tokenizer, stream);
+            }
+
+            @Override
+            protected Reader initReader(String fieldName, Reader reader) {
+                return new ControlCharFilter(reader);
+            }
+        };
+        CharArraySet neverSplitSet = new CharArraySet(neverSplit, false);
+        CharSeqTokenTrieNode neverSplitTree;
+        try (analyzer) {
+            neverSplitTree = CharSeqTokenTrieNode.build(neverSplit, c -> {
+                try (TokenStream ts = analyzer.tokenStream("never_split", c)) {
+                    CharTermAttribute term = ts.addAttribute(CharTermAttribute.class);
+                    ts.reset();
+                    List<String> tokens = new ArrayList<>();
+                    while (ts.incrementToken()) {
+                        tokens.add(term.toString());
+                    }
+                    return tokens;
+                }
+            });
+        }
+        return new BasicTokenFilter(input, neverSplitTree, neverSplitSet, isStripAccents, isTokenizeCjkChars);
+    }
+
+    public BasicTokenFilter(
+        TokenStream input,
+        CharSeqTokenTrieNode neverSplit,
+        CharArraySet neverSplitSet,
+        boolean isStripAccents,
+        boolean isTokenizeCjkChars
+    ) {
+        super(input);
+        this.neverSplit = neverSplit;
+        this.neverSplitSet = neverSplitSet;
+        this.tokens = new LinkedList<>();
+        this.isStripAccents = isStripAccents;
+        this.normalizer = Normalizer2.getNFDInstance();
+        this.splitOn = cp -> (isTokenizeCjkChars && isCjkChar(cp)) || isPunctuationMark(cp);
+    }
+
+    @Override
+    public void reset() throws IOException {
+        super.reset();
+        tokens.clear();
+        accentBuffer.setLength(0);
+        current = null;
+    }
+
+    @Override
+    public boolean incrementToken() throws IOException {
+        if (tokens.isEmpty() == false) {
+            assert current != null;
+            DelimitedToken token = tokens.removeFirst();
+            restoreState(current); // keep all other attributes untouched
+            termAtt.setEmpty().append(token.charSequence());
+            offsetAtt.setOffset(token.startOffset(), token.endOffset());
+            return true;
+        }
+        current = null; // not really needed, but for safety
+        if (input.incrementToken()) {
+            if (isStripAccents) {
+                stripAccent();
+            }
+            if (neverSplitSet.contains(termAtt)) {
+                return true;
+            }
+            int startOffset = offsetAtt.startOffset();
+            // split punctuation and maybe cjk chars!!!
+            LinkedList<DelimitedToken> splits = new LinkedList<>();
```

Comment:
This method is too long. Could we extract this bit all the way to line 169 into a `splitPunctuation` sort of method?

Reply:
for sure